shapefile-reader.js

```diff
@@ -19,35 +22,91 @@ function ShapefileStream (files, options) {
   }
   if (!files.shp) throw new Error('No shp property is set')
 
-  this.header = null
+  this.headers = null
 
   // Create readers
+  this._files = files
   this._shpReader = new ShpReader(files.shp)
+  if (files.dbf) this._dbfReader = new DbfReader(files.dbf)
 
   // State
   this._inBody = false
 }
 
+ShapefileStream.prototype._readHeader = function (callback) {
+  const self = this
+
+  // read character encoding from .cpg
+  if (this._files.hasOwnProperty('cpg')) {
+    this._files.cpg.pipe(endpoint(function (err, buffer) {
+      if (err) return callback(err)
+
+      self._dbfReader.setEncoding(buffer.toString('ascii'))
+      phase2()
+    }))
+  } else {
+    phase2()
+  }
+
+  // read header from .shp (contains shapes) and .bdf (contains properties)
+  function phase2 () {
+    // always read .shp
+    const jobs = {
+      shp: (done) => self._shpReader.readHeader(done)
+    }
+
+    // if .dbf exists, read that too
+    if (self._dbfReader) {
+      jobs.bdf = (done) => self._dbfReader.readHeader(done)
+    }
+
+    // read headers
+    async.parallel(jobs, callback)
+  }
+}
+
+ShapefileStream.prototype._readRecord = function (callback) {
+  if (this._dbfReader) {
+    async.parallel({
+      shp: (done) => this._shpReader.readRecord(done),
+      dbf: (done) => this._dbfReader.readRecord(done)
+    }, function (err, record) {
+      if (err) return callback(err)
+
+      if (record.shp === null && record.dbf === null) {
+        callback(null, null)
+      } else if (record.shp !== null && record.dbf !== null) {
+        record.shp.properties = record.dbf
```

Comment:
Are they simply interleaved like this? There's no indices that have to match or anything?